Signals (#2)
temporalio/edu-interacting-with-workflows-typescript: signaling-your-workflows/common-issues-with-signals.md

# Common Problems with Signals and Their Solutions

> This chapter will discuss common problems people may encounter when using Signals, and their workarounds.

Author: I agree, I have changed mentions of solutions to workarounds

## Limiting Signals

If a Workflow Execution has 2,000 incomplete concurrent tasks (Activities, Signals, Child Workflows, or external Workflow Cancellations), additional Commands of those types (e.g., `ScheduleActivityTask`, `SignalExternalWorkflowExecution`, `StartChildWorkflowExecution`, `RequestCancelExternalWorkflowExecution`) will fail to be applied to that Workflow Execution.

> Furthermore, the total number of Signals received per Workflow Execution cannot exceed 10,000. After that limit is reached, no more Signals will be processed for that Workflow Execution. Also, keep in mind that sending a large number of Signals can also enlarge the Event History size of a Workflow, potentially impacting performance. This is where a concept called [`Continue-As-New`](https://docs.temporal.io workflows#continue-as-new) will be helpful, which you will see later on this page.

Author: Fixed and also removed backticks from Continue-As-New in other spots

> Consider the `pizzaWorkflow` example. Imagine that this Workflow receives a massive influx of Signals (e.g., it receives a Signal every time it receives an order) during a promotional event. If this rate is extremely high, you might quickly approach the limit of concurrent actions or total Signals allowed. To mitigate this, instead of sending a Signal for every individual order, you can batch orders. For example, for every ten orders that are complete, you can take the payload of all ten orders and send a single Signal. This reduces the total number of Signals sent and helps in staying within Temporal's limits.

Author: Thanks for pointing that out! I used a more general example of an online store that uses a Workflow that processes orders.

## Unhandled Command

In scenarios where a Workflow receives a high volume of Signals in a short period of time, it might not process all of them before attempting to complete or transition to another state. This situation leads to an `UnhandledCommand` error because Temporal expects all Signals (new Events) to be addressed before the Workflow can proceed or close.

> In many cases with the TypeScript SDK, there is no need to worry about the "Unhandled Command" error, as it is a transient error. Temporal's server will initiate a new Workflow task that includes the unhandled Signal, and the Worker will process it in the next execution cycle. However, problems arise when a Workflow constantly receives Signals at a high rate, for example, a Signal every second. This can lead to a buildup of Unhandled Commands and an exponentially growing Event History, resulting in performance degradation. As mentioned above, batching Signals would be a good workaround for this scenario.

Author: fixed

> In many cases with the TypeScript SDK, there is no need to worry about the "Unhandled Command" error, as it is a transient error. Temporal's server will initiate a new Workflow task that includes the unhandled Signal, and the Worker will process it in the next execution cycle. However, problems arise when a Workflow constantly receives Signals at a high rate, for example, a Signal every second. This can lead to a buildup of Unhandled Commands and an exponentially growing Event History, resulting in performance degradation. As mentioned above, batching Signals would be a good workaround for this scenario.

Author: From what I know,  it sounds like the Go SDK is the only SDK that needs to worry about the Unhandled Command and will need to properly drain the Signal channels. This does not seem like it applies to any other SDK, but that needs to be verified!

As per your comment, I agree after reading yours that it's confusing and I fixed it to emphasize that it is the rate of the commands that can be harmful.

## Blocking Inside a Signal Handler

> When a Signal handler within a Workflow performs a blocking operation, such as executing an Activity or initiating asynchronous work, it can impede the Workflow's ability to process incoming Signals. This is particularly problematic if the Workflow receives multiple Signals in rapid succession which will lead to a significant degradation in performance.

Author: fixed

> To address these challenges, here is a pattern that manages Signals in a synchronous manner. This approach involves queuing incoming Signals and processing them sequentially one at a time to avoid blocking the Workflow's ability to receive and handle additional Signals.

Author: removed!

```typescript
// Make sure Request is serializable to JSON (ie. no function, no promises, etc)
type Request = { ... }

// Entity workflow pattern with serialization of request
// (ie. only one request is processed at a time)
export async function myWorkflow(requests: Request[] = []): Promise<void> {
  setHandler(mySignal, (input: Request) => {
    requests.push(input);

>     // Don't await here. Otherwise, the Workflow may complete before the promise completes.

Author: Split

  });

  while (!workflowInfo().continueAsNewSuggested) {
		const timeSinceStart = Date.now() - workflowInfo().runStartTime.getTime();
    const shouldProcessMore = await condition(() => requests.length > 0, ms('24h') - timeSinceStart);
    if (!shouldProcessMore) break;

    const request = requests.shift();

    // Process request as appropriate
    await handleSingleRequest(request);
  }

>   // Huge histories are bad for performance, so we switch to a new workflow execution whenever history grows over 2000 events. When that happens, we forward any outstanding requests to the next execution.

Author: Done

  await continueAsNew(requests);
}

function handleSingleRequest(request: Request): Promise<void> {
  // It's ok to await here
}
```

In this pattern:

- Signals are queued instead of being processed immediately. This ensures that the Workflow remains responsive and can handle new Signals as they arrive.
- The queued Signals are processed one at a time, maintaining sequential order and avoiding the risks associated with asynchronous handlers.

> - The Workflow uses the [`Continue-As-New`](https://docs.temporal.io/workflows#continue-as-new) feature to maintain performance by avoiding a large Event History. `Continue-As-New` prevents issues with the Event History's size limitations by spawning a new Workflow with a fresh Event History. The new Workflow Execution will also have the same Workflow ID and the appropriate parameters are passed in. However, it will have a different Run ID.

Author: Fixed both

Important Considerations:

- **Handling Asynchronous Work**: This pattern ensures that all asynchronous work is collected before continuing to the next execution while ensuring that all Signals get handled.

> - **Payload Size Limitations**: This pattern assumes that the Signal payloads are small enough to be carried over in the `Continue-As-New` input payload.

Author: Fixed

> - **Processing Wait Time**: The 24-hour waiting period in this pattern is not a mandatory duration but rather an example for Workflows that don't have a natural endpoint or are long-running. The 24-hour wait in the condition function in the exmample prevents the Workflow from being in an endless wait state if no new Signals are received. Regularly continuing as new (every 24 hours, for example) helps in resetting the Workflow's history and avoiding performance degradation that can occur with excessively large histories. Here are some more time considerations:

Author: Whoops - fixed

# Common Problems with Signals and Their Solutions

This chapter will discuss common problems people may encounter when using Signals, and their workarounds.

## Limiting Signals

> If a Workflow Execution has 2,000 incomplete concurrent Activities, Signals, Child Workflows, or external Workflow Cancellation requests, additional Commands of that type (e.g., `ScheduleActivityTask`, `SignalExternalWorkflowExecution`, `StartChildWorkflowExecution`, `RequestCancelExternalWorkflowExecution`) will fail to be applied to that Workflow Execution.

Author: Yeah yours is less word-soup

> Furthermore, the total number of Signals received per Workflow Execution cannot exceed 10,000. After that limit is reached, no more Signals will be processed for that Workflow Execution. Also, keep in mind that sending a large number of Signals can also enlarge the Event History size of a Workflow, potentially impacting performance.

Reviewer: Is this a good time to mention Continue-As-New, even if it's a link to the Docs?

Author: Added

Consider the `pizzaWorkflow` example. Imagine that this Workflow receives a massive influx of Signals (e.g., it receives a Signal every time it receives an order) during a promotional event. If this rate is extremely high, you might quickly approach the limit of concurrent actions or total Signals allowed. To mitigate this, instead of sending a Signal for every individual order, you can batch orders (Signals). For example, for every ten orders that are complete, you can take the payload of all ten orders and send a single Signal. This reduces the total number of Signals sent and helps in staying within Temporal's limits.

## Unhandled Command

In scenarios where a Workflow receives a high volume of Signals in a short period of time, it might not process all of them before attempting to complete or transition to another state. This situation leads to an `UnhandledCommand` error because Temporal expects all Signals (new Events) to be addressed before the Workflow can proceed or close.

> In the Go SDK, Workflows use channels to handle Signals. It's crucial to ensure that the Signal channel is adequately drained before the Workflow attempts to complete or transition. Not doing so can result in an `Unhandled Command` error because the Workflow might close while there are still Signals in the channel that haven’t been processed. Note that other SDK's do not follow this draining pattern since they do not have these Signal channels and therefore do not have to handle draining.

Reviewer: This seems scary to learn about in passing (though I appreciate you thinking ahead to something we'll have to document in Go :) ) -- draining channels isn't a Temporal concept that most people are going to be thinking about at this point. I would probably remove it from this course and only address it in Go specifically if it's a fairly rare and complicated edge case.

Author: Okay removed! Makes it shorter too which is nice!

> In the Go SDK, Workflows use channels to handle Signals. It's crucial to ensure that the Signal channel is adequately drained before the Workflow attempts to complete or transition. Not doing so can result in an `Unhandled Command` error because the Workflow might close while there are still Signals in the channel that haven’t been processed. Note that other SDK's do not follow this draining pattern since they do not have these Signal channels and therefore do not have to handle draining.

Author: Okay removed! Makes it shorter too which is nice!

> In the Go SDK, Workflows use channels to handle Signals. It's crucial to ensure that the Signal channel is adequately drained before the Workflow attempts to complete or transition. Not doing so can result in an `Unhandled Command` error because the Workflow might close while there are still Signals in the channel that haven’t been processed. Note that other SDK's do not follow this draining pattern since they do not have these Signal channels and therefore do not have to handle draining.

Author: Removed!

> In many cases with the TypeScript SDK, there is no need to worry about the "Unhandled Command" error, as it is a transient error. Temporal's server will initiate a new Workflow task that includes the unhandled Signal, and the Worker will process it in the next execution cycle. However, problems arise when a Workflow constantly receives Signals at a high rate, for example, a Signal every second. This can lead to a buildup of Unhandled Commands and an exponentially growing Event History. Some customers will experience downtime due to this. As mentioned above, batching Signals would be a good workaround for this scenario.

Reviewer: "Some customers will experience downtime due to this" feels too hand-wavy and potentially worrying to me -- I wonder if we should just remove this sentence, since a) it's something we could theoretically improve caching and scheduling for on our end in the future, and b) most of these courses aren't directly addressing uptime of a hosted service.

Author: Referred to performance degradation

## Blocking Inside a Signal Handler

When a Signal handler within a Workflow performs a blocking operation, such as executing an Activity or initiating asynchronous work, it can impede the Workflow's ability to process incoming Signals. This is particularly problematic if the Workflow receives multiple Signals in rapid succession which will lead to a significant degradation in performance.

> Asynchronous Signal handlers can also be risky as they may initiate operations that remain incomplete when the Workflow execution concludes. Imagine a scenario where a Workflow is about to complete its execution, but it receives a Signal that triggers asynchronous work, like calling an Activity or starting a timer. If the Workflow ends before this asynchronous task completes, the task will remain unfinished.

Author: Well I mention signal handlers earlier than this ( I think you were just reading this PR top to bottom maybe :)! But the intro to signal handlers comes before this chapter. I'll change this to 'Signal handlers that perform asynchronous work can....'

> Asynchronous Signal handlers can also be risky as they may initiate operations that remain incomplete when the Workflow execution concludes. Imagine a scenario where a Workflow is about to complete its execution, but it receives a Signal that triggers asynchronous work, like calling an Activity or starting a timer. If the Workflow ends before this asynchronous task completes, the task will remain unfinished.

Author: clarified

To address these challenges, here is a pattern that manages Signals in a synchronous manner. This approach involves queuing incoming Signals and processing them sequentially one at a time to avoid blocking the Workflow's ability to receive and handle additional Signals.

```typescript
// Make sure Request is serializable to JSON (ie. no function, no promises, etc)
type Request = { ... }

// Entity workflow pattern with serialization of request
// (ie. only one request is processed at a time)
export async function myWorkflow(requests: Request[] = []): Promise<void> {
  setHandler(mySignal, (input: Request) => {
    requests.push(input);
    // Don't await here. Otherwise, the Workflow may complete before the promise completes.
  });

  while (!workflowInfo().continueAsNewSuggested) {
		const timeSinceStart = Date.now() - workflowInfo().runStartTime.getTime();

>     const shouldProcessMore = await condition(() => requests.length > 0, ms('24h') - timeSinceStart);

Reviewer: This part feels a little arbitrary -- why is it 24h?

Author: Added explanation in the prose!

    if (!shouldProcessMore) break;

    const request = requests.shift();

    // Process request as appropriate
    await handleSingleRequest(request);
  }

>   // Huge histories are bad for performance, so we switch to a new workflow execution whenever history grows over 2000 events. When that happens, we forward any outstanding requests to the next execution.

Author: recommended by the docs here: https://docs.temporal.io/cloud/limits#per-workflow-execution-concurrency-limits

  await continueAsNew(requests);
}

function handleSingleRequest(request: Request): Promise<void> {
  // It's ok to await here
}
```

In this pattern:

- Signals are queued instead of being processed immediately. This ensures that the Workflow remains responsive and can handle new Signals as they arrive.
- The queued signals are processed one at a time, maintaining sequential order and avoiding the risks associated with asynchronous handlers.
- The Workflow uses the [`Continue-As-New`](https://docs.temporal.io workflows#continue-as-new) feature to maintain performance by avoiding a large Event History. `Continue-As-New` prevents issues with the Event History's size limitations by spawning a new Workflow with a fresh Event History. The new Workflow Execution will also have the same Workflow ID and the appropriate parameters are passed in. However, it will have a different Run ID.

Important Considerations:

> - **Handling Asynchronous Work**: If Signals trigger asynchronous tasks, it's crucial to ensure these tasks are completed before the Workflow attempts to close or `Continue-As-New`.

Reviewer 2: I'm confused here. Are you saying that you could signal the execution of an Activity that is executed asynchronously and the Workflow could close before the Activity completed? If so, isn't that more of an error on the users part to not await the result of the Activity?

Author: Fixed

# Common Problems with Signals and Their Workarounds

This chapter will discuss common problems people may encounter when using Signals, and their workarounds.

> ## Limiting Signals

Reviewer 3: There are multiple issues with this section. Let me clarify the facts, and I'll let you consider how this impact the content of this section.

There are essentially three types of constraints on signals:

1. **Constraints on _sending_ signals from a workflow to another workflow (i.e. using the `SignalExternalWorkflowExecution` workflow command).** The server imposes a limit of 2000 pending outgoing signal requests per workflow execution (ie. the limit is on the workflow that _sends_ signal, not the workflow that receives it). Note that an outgoing signal is "pending" for the time it takes for the cluster to write the corresponding signal event to the other workflow history's and/or mutable state, which should be very fast (i.e. assuming a healthy cluster, all outgoing signals will very likely no longer be in pending state by the time the next WFT get dispatched). Therefore, the only practical impact of this limitation on sending signals is that a single workflow can't send more than 2000 signals at once (ie. from a practical perspective, "at once" usually means "from a single WFT"). If one wants to send a large number of outgoing signals from a workflow, they will need to write their workflow logic to send them as batch of no more than 2000 outgoing signals, making a short pause between each batch. Exceeding that limit will result in a Workflow Task failure.

2. **Constraints on _receiving_ signals**. A single workflow execution can't receive more than 10000 signals in its whole life span, and its history can't exceed 51200 events or 50 MB (...and should ideally remain below 10K events / 10 MB). Exceeding these limits results in the cluster failing the send a signal request (i.e. the call to `workflowHandle.signal(...)` will throw an error; this is true both using the Client API and Workflow commands; similar error for CLI). That means that the sender _knows_ either the signal has been recorded to the target workflow (i.e. overflowing signals are not silently dropped or ignored).

    The server will also fail a send signal request (i.e. the call to `workflowHandle.signal(...)` will throw an error) if the receiving workflow execution has previously attempted to complete or do continue-as-new but failed to do so due to unhandled commands, AND there is an in-flight WFT (i.e. the WFT has been started and not yet completed). This restriction has been added to avoid scenarios where a workflow execution is unable to complete due to repeated "unhandled command" errors.

3. **Limits related to buffered signals**. While there is an in-flight workflow task (for clarity: this is about the receiving workflow), incoming signals are buffered by the server into the mutable state (i.e. they are not immediately written to the workflow history; note that the mutable state is also persistent, so there is no risk toward the safety of these buffered signals; they will get appended to the workflow history once the in-flight WFT completes/times out). Now, there are two restrictions regarding to buffered signals. First, the workflow can't complete nor continue-as-new if there are buffered signals; in that case, the server will refuse the Complete or CAN command with the infamous Unhandled Command error. Also, the server will not accept to buffer more than 100 incoming signals (or a total of 2 MB) for a given workflow execution; if the cluster receive Signal request in excess to these numbers, then the cluster will fail the in-flight WFT, with "WORKFLOW_TASK_FAILED_CAUSE_FORCE_CLOSE_COMMAND" (not sure the exact message users will see).

    Both of these errors are transient and can _generally_ be ignored. The server will simply emit a new WFT, which will include all buffered signals, and a worker will reprocess that. As explained before, there is no longer a risk for recurring Unhandled Command. There is however a risk of recurring "WORKFLOW_TASK_FAILED_CAUSE_FORCE_CLOSE_COMMAND" if the workflow is receiving signals at a high frequency (>100 incoming signals in the time it takes to completes a WFT round trip). In practice, however, this is not frequent.

Furthermore, the total number of Signals received per Workflow Execution cannot exceed 10,000. After that limit is reached, no more Signals will be processed for that Workflow Execution. Also, keep in mind that sending a large number of Signals can also enlarge the Event History size of a Workflow, potentially impacting performance. This is where a concept called [Continue-As-New](https://docs.temporal.io/workflows#continue-as-new) will be helpful, which you will see later on this page.

Consider the example of an online store that uses a Workflow that processes orders. Imagine that this Workflow receives a massive influx of Signals (e.g., it receives a Signal every time it receives an order) during a promotional event. If this rate is extremely high, you might quickly approach the limit of concurrent actions or total Signals allowed. To mitigate this, instead of sending a Signal for every individual order, you can batch orders. For example, for every ten orders that are complete, you can take the payload of all ten orders and send a single Signal. This reduces the total number of Signals sent and helps in staying within Temporal's limits.

> There will be an example of how to do so in the next chapter.

Reviewer 3: ```suggestion
There will be an example of how to do so later in this chapter.
```

## Unhandled Command

In scenarios where a Workflow receives a high volume of Signals in a short period of time, it might not process all of them before attempting to complete or transition to another state. This situation leads to an `UnhandledCommand` error because Temporal expects all Signals (new Events) to be addressed before the Workflow can proceed or close.

In many cases with the TypeScript SDK, it is acceptable to ignore the ocassional "Unhandled Command" error, as it is a transient error. Temporal's server will initiate a new Workflow Task that includes the unhandled Signal, and the Worker will process it in the next execution cycle. However, problems arise when a Workflow constantly receives Signals at a high rate, for example, a Signal every second. This can lead to a buildup of unhandled commands and an exponentially growing Event History, resulting in performance degradation. As mentioned above, batching Signals would be a good workaround for this scenario. To emphasize, it is not that some "Unhandled Command" errors are important and others are not - what matters is the rate of them. 

> ## Blocking Inside a Signal Handler

Reviewer 3: I think the warning regarding async signal handler deserves its own subtitle, and "Blocking" would not be the proper word for that.

As for introduction of the entity pattern, I think it should be moved to a different page. Presenting this as a "workaround" doesn't feel appropriate to me. We commonly call this the Entity Workflow Pattern, and it is a very common and strong architectural pattern for development of signal-based workflows, including "state machines as a workflow", and "actor as a workflow" things.

> When a Signal handler within a Workflow performs a blocking operation, such as executing an Activity, it can impede the Workflow's ability to process incoming Signals. This is particularly problematic if the Workflow receives multiple Signals in rapid succession which will lead to a significant degradation in performance.

Reviewer 3: That's not true in TS, nor in any other SDK actually, with a minor exception in Java SDK that may happen if the worker runs out of worker threads.

Actually, it's quite the opposite: if multiple signals come in and execute asynchronous stuff in your signal handler, then these signals effectively gets executed concurrently, and therefore, you loose the predictability of operation ordering. The SDK guarantee that they will always reexecute in the same order (otherwise that would break determinism), but it is possible for the signal handler to complete processing of signal 2 before it completes processing of signal 1 (which came first). Depending on the use case, this may or may not be a problem.

Another thing to be careful with is that a workflow completes when it reaches the end of the workflow's main function, or as soon as it calls `continueAsNew`. Signal handlers that started processing a signal but blocked on an async operation will not prevent completion of the workflow, and pending async operations (timer, activities, and child workflow unless they are configured to get abandoned) will simply get cancelled at that point. This is a serious drawback for most use case.

These are the two reasons for the pattern you demonstrate below.

> To address these challenges, here is a pattern that manages Signals in a synchronous manner. This approach involves queuing incoming Signals and processing them to avoid blocking the Workflow's ability to receive and handle additional Signals.

Reviewer 3: ```suggestion
To address these challenges, here is a pattern that manages Signals in a synchronous manner. This approach involves internal queuing of incoming Signals, which are then processed one at a time, therefore ensuring a well defined execution order and predictable behavior.
```
